File: atlhyper_web/src/api/event.ts
/**
 * Event API
 *
 * 后端已返回 camelCase 响应，前端直接使用
 */

import { get } from "./request";
import type { EventOverview, EventLog, EventCards } from "@/types/cluster";

// ============================================================
// 查询参数类型
// ============================================================

interface EventListParams {
  cluster_id: string;
  namespace?: string;
  type?: string; // Normal, Warning
  reason?: string;
  involved_kind?: string;
  involved_name?: string;
  since?: string; // RFC3339 格式
  limit?: number;
  offset?: number;
  source?: string;
}

// ============================================================
// 响应类型（后端已返回 camelCase）
// ============================================================

interface EventListResponse {
  events: EventLog[];
  total: number;
  source?: string;
  limit?: number;
  offset?: number;
}

// ============================================================
// 聚合工具（从事件列表生成概览统计）
// ============================================================

function aggregateOverview(events: EventLog[]): EventOverview {
  const kindsSet = new Set<string>();
  const categoriesSet = new Set<string>();
  let warning = 0;
  let error = 0;
  let info = 0;

  for (const e of events) {
    if (e.kind) kindsSet.add(e.kind);
    if (e.reason) categoriesSet.add(e.reason);

    const severity = e.severity?.toLowerCase() || "";
    if (severity === "warning") {
      warning++;
    } else if (severity === "error") {
      error++;
    } else {
      info++;
    }
  }

  return {
    cards: {
      totalAlerts: warning + error,
      totalEvents: events.length,
      warning,
      error,
      info,
      kindsCount: kindsSet.size,
      categoriesCount: categoriesSet.size,
    },
    rows: events,
  };
}

// ============================================================
// API 方法
// ============================================================

/**
 * 获取事件列表（实时，从 DataHub）
 * GET /api/v2/events?cluster_id=xxx&namespace=xxx&type=xxx
 */
export function getEventList(params: EventListParams) {
  return get<EventListResponse>("/api/v2/events", params);
}

/**
 * 获取告警列表（历史，从数据库）
 * GET /api/v2/events?cluster_id=xxx&source=history
 */
export function getAlertList(params: EventListParams) {
  return get<EventListResponse>("/api/v2/events", {
    ...params,
    source: "history",
  });
}

/**
 * 获取事件概览（实时，从 DataHub）
 */
export async function getEventOverview(data: { ClusterID: string; Namespace?: string; Type?: string }) {
  const response = await getEventList({
    cluster_id: data.ClusterID,
    namespace: data.Namespace,
    type: data.Type,
  });

  const events = response.data.events || [];

  return {
    ...response,
    data: {
      data: aggregateOverview(events),
    },
  };
}

/**
 * 获取告警概览（历史，从数据库）
 */
export async function getAlertOverview(data: { ClusterID: string }) {
  const response = await getAlertList({
    cluster_id: data.ClusterID,
  });

  const events = response.data.events || [];

  return {
    ...response,
    data: {
      data: aggregateOverview(events),
    },
  };
}

/** @deprecated 使用 getAlertOverview 替代 */
export function getEventLogs(data: { ClusterID: string; Namespace?: string; Type?: string }) {
  return getAlertOverview(data);
}

File: atlhyper_master_v2/model/convert/event.go
// atlhyper_master_v2/model/convert/event.go
// model_v2.Event / database.ClusterEvent → model.EventLog 转换函数
package convert

import (
	"time"

	"AtlHyper/atlhyper_master_v2/database"
	"AtlHyper/atlhyper_master_v2/model"
	"AtlHyper/model_v2"
)

// EventLog 转换单个 model_v2.Event 为 Web API 响应
func EventLog(src *model_v2.Event, clusterID string) model.EventLog {
	kind := src.InvolvedObject.Kind
	if kind == "" {
		kind = "Event"
	}

	name := src.InvolvedObject.Name
	if name == "" {
		name = src.Name
	}

	return model.EventLog{
		ClusterID: clusterID,
		Category:  src.Reason,
		EventTime: formatTime(src.LastTimestamp),
		Kind:      kind,
		Message:   src.Message,
		Name:      name,
		Namespace: src.Namespace,
		Node:      src.Source,
		Reason:    src.Reason,
		Severity:  mapSeverity(src.Type),
		Time:      formatTime(src.FirstTimestamp),
	}
}

// EventLogs 转换多个 model_v2.Event
func EventLogs(src []model_v2.Event, clusterID string) []model.EventLog {
	if src == nil {
		return []model.EventLog{}
	}
	result := make([]model.EventLog, len(src))
	for i := range src {
		result[i] = EventLog(&src[i], clusterID)
	}
	return result
}

// EventLogFromDB 转换数据库事件记录为 Web API 响应
func EventLogFromDB(src *database.ClusterEvent, clusterID string) model.EventLog {
	return model.EventLog{
		ClusterID: clusterID,
		Category:  src.Reason,
		EventTime: formatTime(src.LastTimestamp),
		Kind:      src.InvolvedKind,
		Message:   src.Message,
		Name:      src.InvolvedName,
		Namespace: src.Namespace,
		Node:      src.SourceComponent,
		Reason:    src.Reason,
		Severity:  mapSeverity(src.Type),
		Time:      formatTime(src.FirstTimestamp),
	}
}

// EventLogsFromDB 转换多个数据库事件记录
func EventLogsFromDB(src []*database.ClusterEvent, clusterID string) []model.EventLog {
	if src == nil {
		return []model.EventLog{}
	}
	result := make([]model.EventLog, len(src))
	for i, e := range src {
		result[i] = EventLogFromDB(e, clusterID)
	}
	return result
}

// EventOverview 将事件列表转换为概览（含统计卡片）
func EventOverview(src []model_v2.Event, clusterID string) model.EventOverview {
	rows := EventLogs(src, clusterID)

	kinds := map[string]struct{}{}
	categories := map[string]struct{}{}
	var warning, errCount, info int

	for _, r := range rows {
		if r.Kind != "" {
			kinds[r.Kind] = struct{}{}
		}
		if r.Reason != "" {
			categories[r.Reason] = struct{}{}
		}
		switch r.Severity {
		case "warning":
			warning++
		case "error":
			errCount++
		default:
			info++
		}
	}

	return model.EventOverview{
		Cards: model.EventCards{
			TotalAlerts:     warning + errCount,
			TotalEvents:     len(rows),
			Warning:         warning,
			Info:            info,
			Error:           errCount,
			CategoriesCount: len(categories),
			KindsCount:      len(kinds),
		},
		Rows: rows,
	}
}

// mapSeverity 映射 K8s Event Type 到 severity
func mapSeverity(eventType string) string {
	switch eventType {
	case "Warning":
		return "warning"
	case "Error":
		return "error"
	default:
		return "info"
	}
}

// formatTime 格式化 time.Time 为 RFC3339 字符串
func formatTime(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339)
}
